test/freenet/client/async/HealingDecisionSupplierTest.java

```diff
@@ -12,31 +12,62 @@ public class HealingDecisionSupplierTest {
   @Test
   public void healingAlwaysAcceptsAKeyAtTheNodeLocation() {
     for (double i : Arrays.asList(1.0, 0.9, 0.8, 0.7, 0.6, 0.5, 0.4, 0.3, 0.2, 0.1, 0.00000001)) {
-      assertThat(String.format("Healing triggers at random value %g", i), HealingDecisionSupplier.shouldHealBlock(0.1, 0.1, i), Matchers.equalTo(true));
+      HealingDecisionSupplier healingDecisionSupplier = new HealingDecisionSupplier(
+          () -> 0.1,
+          () -> true,
+          () -> i);
+      assertThat(
+          String.format("Healing triggers at random value %g", i),
+          healingDecisionSupplier.shouldHeal(0.1),
+          Matchers.equalTo(true));
     }
   }
 
   @Test
   public void healingAccepts70PercentOfKeysInShortDistance() {
     for (double i : Arrays.asList(1.0, 0.9, 0.8, 0.7, 0.6, 0.5, 0.4)) {
-      assertThat(String.format("Healing triggers at random value %g", i), HealingDecisionSupplier.shouldHealBlock(0.1, 0.11, i), Matchers.equalTo(true));
+      HealingDecisionSupplier healingDecisionSupplier = new HealingDecisionSupplier(
+          () -> 0.1,
+          () -> true,
+          () -> i);
+      assertThat(String.format("Healing triggers at random value %g", i),
+          healingDecisionSupplier.shouldHeal(0.11),
+          Matchers.equalTo(true));
     }
   }
 
   @Test
   public void healingAccepts50PercentOfKeysAtTheLimitOfLongDistance() {
     for (double i : Arrays.asList(1.0, 0.9, 0.8, 0.7, 0.6)) {
-      assertThat(String.format("Healing triggers at random value %g", i), HealingDecisionSupplier.shouldHealBlock(0.1, 0.1999, i), Matchers.equalTo(true));
+      HealingDecisionSupplier healingDecisionSupplier = new HealingDecisionSupplier(
+          () -> 0.1,
+          () -> true,
+          () -> i);
+      assertThat(String.format("Healing triggers at random value %g", i),
+          healingDecisionSupplier.shouldHeal(0.1999),
+          Matchers.equalTo(true));
     }
   }
 
   @Test
   public void healingAccepts10PercentOfKeysAtLongDistance() {
```

Comment:
This is the only test that actually tests both scenarios (i.e. `shouldHeal == true` and `shouldHeal == false`). I’m missing those cases for the the other tests as well. 🙂 Also, `isOpennetEnabled` is never false, and the location is always `0.1`. The way the tests are currently written you could probably write an implementation of the HealingDecisionSupplier that is quite broken without the brokenness being detected by the tests.

Reply:
The tests now include inverse tests, too.

Reply:
Opennet is false in the first test. But the location is always 0.1, yes: this is not intended to catch intentional errors, just to catch accidental ones.

But anyway: adding more tests of the same structure doesn’t cost much: https://github.com/hyphanet/fred/pull/874/commits/cb837421717060dc3fd490f0deea27c872797e7c → different node location, and diff in positive and in negative direction.

Reply:
Also a test around zero: https://github.com/hyphanet/fred/pull/874/commits/4b26dbc2760993b96db69efc7549ba4df1a1b065